extensions/amp-story/0.1/page-scaling.js

```diff
@@ -0,0 +1,253 @@
+/**
+ * Copyright 2018 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import {Services} from '../../../src/services';
+import {dev, user} from '../../../src/log';
+import {dict} from '../../../src/utils/object';
+import {isExperimentOn} from '../../../src/experiments';
+import {childElementsByTag, matches} from '../../../src/dom';
+import {px, setImportantStyles} from '../../../src/style';
+import {renderAsElement} from './simple-template';
+import {throttle} from '../../../src/utils/rate-limit';
+import {toArray} from '../../../src/types';
+
+
+/** @private @const {number} */
+const MIN_LAYER_WIDTH_PX = 380;
+
+
+/** @private @const {number} */
+const MAX_LAYER_WIDTH_PX = 520;
+
+
+/** @private @const {string} */
+const SCALING_APPLIED_CLASSNAME = 'i-amphtml-story-scaled';
+
+
+/** @private @const {!./simple-template.ElementDef} */
+const SIZER_TEMPLATE = {
+  tag: 'div',
+  attrs: dict({'class': 'i-amphtml-story-page-sizer'}),
+};
+
+
+/** @struct @typedef {{factor: number, width: number, height: number}} */
+let DimensionsDef;
+
+
+/** @typedef {{scale: function(!Element):!Promise}} */
+let PageScalingServiceInterface;
+
+
+/**
+ * @param {!Element} sizer
+ * @return {!DimensionsDef}
+ */
+function targetDimensionsFor(sizer) {
+  const {width, height} = sizer./*OK*/getBoundingClientRect();
+
+  const ratio = width / height;
+
+  const targetWidth = Math.min(MAX_LAYER_WIDTH_PX,
+      Math.max(width, Math.max(1, ratio) * MIN_LAYER_WIDTH_PX));
+
+  const targetHeight = (targetWidth / ratio);
+
+  const factor = width / targetWidth;
+
+  return {width: targetWidth, height: targetHeight, factor};
+}
+
+
+/**
+ * @param {!Element} page
+ * @param {!DimensionsDef} dimensions
+ */
+function applyScaling(page, dimensions) {
+  const {width, height, factor} = dimensions;
+  // TODO(alanorozco): Calculate relative layer dimensions for cases where
+  // they won't fill the entire page.
+  const style = {
+    'width': px(width),
+    'height': px(height),
+    'zoom': `${factor * 100}%`,
```

Comment:
FWIW, I think you can use decimal zooms (e.g. 0.9) instead of percentage (e.g. 90%), so you may just be able to use `factor` directly.